core/src/test/scala/org/apache/spark/util/UninterruptibleThreadSuite.scala

```diff
@@ -115,6 +116,45 @@ class UninterruptibleThreadSuite extends SparkFunSuite {
     assert(interruptStatusBeforeExit)
   }
 
+  test("no runUninterruptibly") {
+    @volatile var hasInterruptedException = false
+    val t = new UninterruptibleThread("test") {
+      override def run(): Unit = {
+        if (sleep(0)) {
+          hasInterruptedException = true
+        }
+      }
+    }
+    t.interrupt()
+    t.start()
+    t.join()
+    assert(hasInterruptedException === true)
+  }
+
+  test("SPARK-51821 uninterruptibleLock deadlock") {
+    val latch = new CountDownLatch(1)
+    val task = new UninterruptibleThread("task thread") {
+      override def run(): Unit = {
+        val channel = new AbstractInterruptibleChannel() {
+          override def implCloseChannel(): Unit = {
+            begin()
+            latch.countDown()
+            try {
+              Thread.sleep(Long.MaxValue)
+            } catch {
+              case _: InterruptedException => Thread.currentThread().interrupt()
+            }
+          }
+        }
+        channel.close()
+      }
+    }
+    task.start()
+    assert(latch.await(100, TimeUnit.SECONDS), "await timeout")
+    task.interrupt()
```

Comment:
When we reach here, the thread is in the uninterruptible state, and `task.interrupt` should only set `shouldInterruptThread` to true. Looks like this deadlock is by design?

Maybe I didn't understand the full event sequence. How did the test thread become interruptible?

Reply:
The task thread never becomes `uninterruptible`. To enter `uninterruptible` state it should be executed using `runUninterruptibly()`. The deadlock happens when task thread is interrupted. Please see my e-mail on dev list.

Reply:
ok so `UninterruptibleThread` can be used as a normal thread without invoking `runUninterruptibly`?

Reply:
nvm I see the discussion in https://github.com/apache/spark/pull/50594/files#r2055362552

Reply:
More importantly `UninterruptibleThread` are used by `Executors` in  a thread pool, so the same thread (the same instance of `UninterruptibleThread`) can be used as a normal thread and as uninterruptible thread.